VirtualApplication support

rafalmaciag opened the pull request:
Virtual-Application creation support with documentation.

SharpeRAD commented:
Can you explain why you didn't expand the existing [application methods](https://github.com/SharpeRAD/Cake.IIS/blob/master/src/IIS/Manager/Base/BaseSiteManager.cs#L482)? Or use the existing [authentication methods](https://github.com/SharpeRAD/Cake.IIS/blob/master/src/IIS/Manager/Base/BaseSiteManager.cs#L137)?

I'm not against adding functionality to the Addin but it looks like you could have just added "EnabledProtocols" to [ApplicationSettings](https://github.com/SharpeRAD/Cake.IIS/blob/master/src/IIS/Settings/ApplicationSettings.cs) and added the "ApplicationExists" method to the [BaseSiteManager](https://github.com/SharpeRAD/Cake.IIS/blob/master/src/IIS/Manager/Base/BaseSiteManager.cs) / [SiteAliases](https://github.com/SharpeRAD/Cake.IIS/blob/master/src/IIS/Aliases/SiteAliases.cs).

rafalmaciag commented:
Hi,

Good point. I haven't found that while exploring features of the Addin :/
Great to see that, and sorry for my misfortune search!!  So I only need
something like overwrite or ApplicationExists - I'll fix that and get back,
and use current methods :-) Thank you!

2016-07-25 20:23 GMT+02:00 SharpeRAD <notifications@github.com>:

> Can you explain why you didn't expand the existing application methods
> <https://github.com/SharpeRAD/Cake.IIS/blob/master/src/IIS/Manager/Base/BaseSiteManager.cs#L482>?
> Or use the existing authentication methods
> <https://github.com/SharpeRAD/Cake.IIS/blob/master/src/IIS/Manager/Base/BaseSiteManager.cs#L137>
> ?
>
> I'm not against adding functionality to the Addin but it looks like you
> could have just added "EnabledProtocols" to ApplicationSettings
> <https://github.com/SharpeRAD/Cake.IIS/blob/master/src/IIS/Settings/ApplicationSettings.cs>
> and added the "ApplicationExists" method to the BaseSiteManager
> <https://github.com/SharpeRAD/Cake.IIS/blob/master/src/IIS/Manager/Base/BaseSiteManager.cs>
> / SiteAliases
> <https://github.com/SharpeRAD/Cake.IIS/blob/master/src/IIS/Aliases/SiteAliases.cs>
> .
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/SharpeRAD/Cake.IIS/pull/6#issuecomment-235039679>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AMOGcKYQ0mWcrj2T3H5IUQa1nmSw88xHks5qZP8rgaJpZM4JUGID>
> .
>

rafalmaciag commented:
Hi, 
I've deleted the unintended approach. Here are my changes. I've added support for setting up security while adding an Application to webSite. I've also deleted some not important check while removing an Virtual Application from WebSite.

SharpeRAD commented:
Look at the "Files changed" section of this PR to see what effect changing the indentation of a file has on a diff...

Please amend the PR to keep the original indentation so I can easily review your changes.

SharpeRAD commented on src/IIS/Manager/Base/BaseSiteManager.cs:
```diff
@@ -1,624 +1,680 @@
+                throw new Exception("Site '" + settings.SiteName + "' does not exist.");
+            }
+        }
+        private static void ValidateApplication(string webSite, string appName)
```
Why is this a separate function?

rafalmaciag replied:
Because it was used a couple of times.

SharpeRAD commented on src/IIS/Settings/ApplicationSettings.cs:
```diff
@@ -8,43 +10,55 @@ namespace Cake.IIS
+        #region Constructor (1)
+        public ApplicationSettings()
+        {
+            Authentication = new AuthenticationSettings();
```
This means authentication will always be set?

rafalmaciag replied:
Yes. This is the intension. Just like I don't return null when the method returns an array - then it should return empty array. Same here. I want the code to be simple, and so any condition that checks for null is yet another comlication.

SharpeRAD replied:
But not everyone will want to set the authentication of the application.

SharpeRAD commented on src/IIS/Manager/Base/BaseSiteManager.cs:
```diff
@@ -1,624 +1,680 @@
+        /// <param name="settings">The site settings</param>
+        protected void SetAuthentication(SiteSettings settings)
+        {
+            var authenticationSettings = settings.Authentication;
```
Why store this as a local variable?

rafalmaciag replied:
I was trying to unify the flow and use only one method for setting authentication - this variable was reused here and there. Nothing major change here.

SharpeRAD commented on src/IIS/Manager/Base/BaseSiteManager.cs:
```diff
@@ -1,624 +1,680 @@
+        private const string anonymAuthKey = "system.webServer/security/authentication/anonymousAuthentication";
+        private const string basicAuthKey = "system.webServer/security/authentication/basicAuthentication";
+        private const string windowsAuthKey = "system.webServer/security/authentication/windowsAuthentication";
+        protected void SetAppAuthentication(string server, string site, string appPath, AuthenticationSettings authenticationSettings)
```
Why have you separated this into two functions?

rafalmaciag replied:
On my machine it didn't worked out for VirtualApplications. I got an exception, something regarding root section not found.

SharpeRAD commented on src/IIS/Manager/Base/BaseSiteManager.cs:
```diff
@@ -1,624 +1,680 @@
+            return site;
+        }
+
+        protected const string webServerName = "webServer";
```
What was wrong with declaring them inline, the variable name is longer than the word?

SharpeRAD commented on src/IIS/Manager/Base/BaseSiteManager.cs:
```diff
@@ -1,624 +1,680 @@
+            if (authenticationSettings != null)
+            {
+                //Authentication
+                SetSiteAuthentication(server, locationPath, authenticationSettings);
```
Why is SetAuthentication now just a wrapper for SetSiteAuthenticaiton?

rafalmaciag replied:
Yes, I didn't want to break any public usage. And so when I divided that into two: SetSiteAuthentication, SetAppAuthentication then it would not break any code that was relying on SetAuthentication.

SharpeRAD commented:
I genuinely want to accept every PR people send me but I'm struggling with this one...

rafalmaciag commented:
Sure, I understand. My initial requirement was to support Authentication to one level nested virtual applications. This works fine, yet in general I would probably do it differently. I tried not to do too many changes and break anything. For instance the SiteSettings are use for removing the site - why to require name of app pool? So I only would like this VirtApp better support. :)

SharpeRAD commented:
I'm going to close this PR for now. If there's a feature that you feel is missing from IIS applications I'd happily accept another focused PR (eg: this adds feature X).
